.github/workflows/release.yml

```diff
@@ -28,55 +45,88 @@ jobs:
+        run: echo "phylum_release_version=$(poetry version --short)" >> $GITHUB_ENV
+
+      # NOTE: Run the tests for the current active Python version, as a sanity check.
       - name: Run tox via poetry
         run: poetry run tox
 
       - name: Build wheel and source distribution
         run: poetry build -vvv
 
-      - name: Create GitHub release
-        id: github_release
-        uses: softprops/action-gh-release@v0.1.14
-        with:
-          # This check is already filtered down to only 'refs/tags/' and shouldn't "overmatch" for a different ref type.
-          # Look for the post-normalized PEP440 pre-release suffixes: https://peps.python.org/pep-0440/#pre-releases
-          prerelease: ${{ contains(github.ref, 'a') || contains(github.ref, 'b') || contains(github.ref, 'rc') }}
-          generate_release_notes: true
-          token: ${{ secrets.GITHUB_TOKEN }}
-          fail_on_unmatched_files: true
-          files: |
-            dist/*.whl
-            dist/*.tar.gz
-
-      - name: Publish to PyPI
-        run: poetry publish --username __token__ --password ${{ secrets.PYPI_API_TOKEN }}
+      - name: Publish to TestPyPI
+        id: publish_test_pypi
+        run: poetry publish --repository testpypi --username __token__ --password ${{ secrets.TESTPYPI_API_TOKEN }}
+
+      # This step is currently only verifying that the package uploaded to TestPyPI can be installed and run from there.
+      # This would be a good spot to trigger a more holistic E2E or integration (as it were) level testing suite. It is
+      # also possible to use the `Production` environment here as a manual gating function. That is, the required
+      # "reviewers" for that environment would be notified of the release and could check that the package verification
+      # step(s) worked as intended before approving the release to proceed.
+      - name: Verify the package
+        run: >
+          pipx run
+          --index-url https://test.pypi.org/simple/
+          --spec "phylum==${{ env.phylum_release_version }}"
+          --pip-args="--extra-index-url=https://pypi.org/simple/"
+          phylum-init -h
+
+      # This step is needed b/c otherwise the Python Semantic Release `publish` cmd would bump the version a 2nd time.
+      - name: Revert to original version
+        run: poetry version ${{ env.phylum_original_version }}
+
+      - name: Use Python Semantic Release to publish release
+        id: psr_release
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          DEBUG: verbose
+          REPOSITORY_USERNAME: __token__
+          REPOSITORY_PASSWORD: ${{ secrets.PYPI_API_TOKEN }}
+        run: |
+          git config user.name github-actions
+          git config user.email github-actions@github.com
+          if [ ${{ github.event.inputs.prerelease }} = "true" ]; then
+            poetry run semantic-release publish --prerelease
+          else
+            poetry run semantic-release publish
```

Comment:
This includes a push to `main`, correct? I guess there's not much of another option.

Reply:
Yes, it does. It is also the source of an unanswered question: will this commit/push to `main` actually work here? The `GITHUB_TOKEN` exists to allow _some_ things like this, but `main` is a protected branch that requires signed commits. Even with the `github-actions <github-actions@github.com>` commit author set in `pyproject.toml` and these lines in the workflow itself:

```
          git config user.name github-actions
          git config user.email github-actions@github.com
```

...for the task that uses PSR publish, it is not clear what will happen. It might not be until this PR is merged and a release is attempted. Should it fail, there are already some options that might be worth pursuing:

* [How to Sign automatic commits in GitHub Actions](https://github.com/Nautilus-Cyberneering/pygithub/blob/f04df405e46a865d80ba3b5729c9b61312320121/docs/how_to_sign_automatic_commits_in_github_actions.md)
* [Verified commits made easy with GitHub Actions](https://gist.github.com/swinton/03e84635b45c78353b1f71e41007fc7c)